sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +298,190 @@ def reduce(self, word, exclude=None):
+                    accept_states.append(elem)
+
+        # Add new transitions for every state.
+        for state in self.reduction_automaton.states:
+            current_state_name = state
+            current_state_type = self.reduction_automaton.states[state].state_type
+            # Transitions will be modified only when current_state belongs to the proper_perfixes of the new rules.
+            # The rest are ignored if they cannot lead to a dead state after a finite number of transisitons.
+            if current_state_type == "start":
+                for letter in automaton_alphabet:
+                    if letter in self.reduction_automaton.states:
+                        self.reduction_automaton.states[state].add_transition(letter, letter)
+                    else:
+                        self.reduction_automaton.states[state].add_transition(letter, current_state_name)
+            else:
+                # Store subwords of current state in an array.
+                current_state_name_array = []
+                next_sub = current_state_name
+                while len(next_sub):
+                    current_state_name_array.append(next_sub)
+                    next_sub = next_sub.subword(1, len(next_sub))
+                # Check if the transition to any new state in posible.
+                for sub_word in current_state_name_array:
+                    if sub_word in accept_states:
+                        # Add accept states.
+                        if current_state_type == "accept":
+                            for letter in automaton_alphabet:
+                                next = current_state_name*letter
+                                while True:
+                                    if len(next) <= 1:
+                                        if next in self.reduction_automaton.states:
+                                            self.reduction_automaton.states[state].add_transition(letter, next)
+                                        else:
+                                            self.reduction_automaton.states[state].add_transition(letter, 'start')
+                                        break
+                                    else:
+                                        if next in self.reduction_automaton.states:
+                                            self.reduction_automaton.states[state].add_transition(letter, next)
+                                            break
+                                        next = next.subword(1, len(next))
+                        break
+
+        # Add transitions for new states. All symbols used in the automaton are considered here.
+        # Ignore this if `generators` = `automaton_alphabet`.
+        if len(generators) != len(automaton_alphabet):
+            for state in proper_prefixes:
+                for letter in generators:
+                    next = current_state_name*letter
```

Comment:
One of the tests in the build fails with `TypeError: 'str' object cannot be interpreted as an integer` on this line. Is thi because the name of the start state is a string, not a group element or is it something else?

Reply:
Yes, it is because of `start` being a string.

Reply:
Looks like it's not because of `start` being a string. `current_state_name` has to be defined again in this scope. I guess I accidentally deleted that when I made these new changes. Anyways, I've updated it now.